fms/models/gpt_bigcode.py

```diff
@@ -0,0 +1,320 @@
+        )
+
+        self.embedding = nn.Embedding(self.config.src_vocab_size, self.config.emb_dim)
+        self.position_embedding = nn.Embedding(self.config.max_pos, self.config.emb_dim)
+
+        self.dec_norm = nn.LayerNorm(self.config.emb_dim, eps=self.config.ln_eps)
+
+        if self.config.emb_dropout:
+            self.emb_dropout = nn.Dropout(self.config.emb_dropout)
+
+        if self.config.p_dropout:
+            self.dropout = nn.Dropout(self.config.p_dropout)
+
+    def forward(
+        self,
+        x: torch.LongTensor,
+        mask: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        past_key_value_states: Optional[
+            Tuple[
+                torch.FloatTensor,
+            ]
+        ] = None,
+        use_cache: bool = False,
+        attn_algorithm: Optional[str] = None,
+    ):
+        # Embed the given vocabulary indices using the given attention mask, with pre-/post-norm and dropout as specified
+        # x_in: batch_size x seq_len
+        # mask: batch_size x seq_len x seq_len
+        # bias: nheads x seq_len x seq_len
+
+        qlen = x.size(1)
+        klen = x.size(1)
+
+        if past_key_value_states is None or len(past_key_value_states) == 0:
+            past_key_value_states = [None for _ in range(len(self.layers))]
+
+        # if we are using the cache, the key length needs to be extended with the past keys length
+        if use_cache and past_key_value_states[0] is not None:
+            klen += past_key_value_states[0][0].size(-2)
+
+        # if mask is none, we need to compute mask
+        is_causal_mask = False
+        if mask is None:
+            if x is None:
+                raise ValueError("cannot create a mask when x is None")
+            # we are caching and can assume all 1s in the mask
+            if use_cache and klen != 1 and qlen == 1:
+                # b x h x qlen x kvlen
+                mask = torch.ones(qlen, klen, device=x.device)
+            else:
+                pad_id: int = self.config.pad_id
+                is_pad: torch.BoolTensor = x == pad_id
+                mask: torch.BoolTensor = is_pad.unsqueeze(-1) == is_pad.unsqueeze(-2)
+                mask = mask.tril(diagonal=0)
+
+        x_emb = self.embedding(x)
+
+        # if pad_id exists
+        #   is_pad will be a BoolTensor
+        #   otherwise pad_id will not be taken into account
+        if self.config.pad_id is None:
+            is_pad = torch.zeros_like(x, dtype=bool, device=x.device)
+        else:
+            is_pad = x == self.config.pad_id
+
+        if position_ids is None:
+            position_ids = (~is_pad).cumsum(1)
+
+            # Compute position_ids based on cache config
+            if use_cache and past_key_value_states[0] is not None:
+                position_ids += past_key_value_states[0][0].size(-2)
+
+            # correct for pads if a pad_id exists
```

Comment:
the idea with constructing position_ids from `(~is_pad).cumsum` is that I don't think you still have to remove pads or clamp anymore.

Reply:
For the input `torch.arange(0, 16).unsqueeze(0)` we will end up getting position_ids [0, 1, ..., 15]. In this case, pad_id=0, so what we really want for position ids is [0, 0, 1, 2, ..., 14]. Unless we add the the below lines, we won't get that.

Reply:
We do still need the clamp, because `(~is_pad).cumsum()` is 1-indexed whereas we need 0-indexed. So really we need `(~is_pad).cumsum().sub(1).clamp(min=0)`, but that's still simpler than what we have now.

Reply:
updated in latest commit